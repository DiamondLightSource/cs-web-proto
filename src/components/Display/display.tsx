import React from "react";
import PropTypes from "prop-types";

import {
  Component,
  Widget,
  WidgetPropType,
  InferWidgetProps
} from "../Widget/widget";

const DisplayProps = {
  children: PropTypes.node
};

// Generic display widget to put other things inside
const DisplayComponent = (
  props: InferWidgetProps<typeof DisplayProps> & Component
): JSX.Element => (
  <div
    style={{ position: "relative", boxSizing: "border-box", ...props.style }}
  >
    {props.children}
  </div>
);

const DisplayWidgetProps = {
  ...DisplayProps,
  ...WidgetPropType
};

export const Display = (
  props: InferWidgetProps<typeof DisplayWidgetProps>
): JSX.Element => <Widget baseWidget={DisplayComponent} {...props} />;

Display.propTypes = DisplayWidgetProps;
