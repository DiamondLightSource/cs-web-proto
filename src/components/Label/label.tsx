import React from "react";

import classes from "./label.module.css";
import { Component, Widget, WidgetPropType } from "../Widget/widget";
import { registerWidget } from "../register";
import { BoolPropOpt, StringOrNumProp, InferWidgetProps } from "../propTypes";

const LabelProps = {
  text: StringOrNumProp,
  visible: BoolPropOpt,
  transparent: BoolPropOpt
};

export const LabelComponent = (
  props: InferWidgetProps<typeof LabelProps> & Component
): JSX.Element => {
  const style: any = { ...props.style };
  if (props.visible !== undefined && !props.visible) {
    style["visibility"] = "hidden";
  }
  if (props.transparent !== undefined && props.transparent) {
    style["backgroundColor"] = "transparent";
  }
  // Simple component to display text - defaults to black text and dark grey background
  return (
    <div className={`Label ${classes.Label}`} style={style}>
      {props.text}
    </div>
  );
};

const LabelWidgetProps = {
  ...LabelProps,
  ...WidgetPropType
};

export const Label = (
  props: InferWidgetProps<typeof LabelWidgetProps>
): JSX.Element => <Widget baseWidget={LabelComponent} {...props} />;

registerWidget(Label, LabelWidgetProps, "label");
