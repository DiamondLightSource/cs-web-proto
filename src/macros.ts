import { MacroMap } from "./redux/csState";

function interpolate(
  str: string,
  substitutions: { [substitution: string]: string }
): string {
  const requiredMappings: string[] = [];

  const regexp = /\${(.*?)}/g;
  let result = null;
  while ((result = regexp.exec(str))) {
    requiredMappings.push(result[1]);
  }
  const missingMappings = requiredMappings.filter(
    (x: string): boolean => substitutions[x] === undefined
  );

  /* Allow missing macros to go unresolved. */
  for (var missing of missingMappings) {
    substitutions[missing] = "${" + missing + "}";
  }

  return str.replace(/\${(.*?)}/g, (x, g): string => substitutions[g]);
}

export function resolveMacros(
  unresolvedPvName: string,
  macroMap: MacroMap
): string {
  const resolved = interpolate(unresolvedPvName, macroMap);
  return resolved;
}
