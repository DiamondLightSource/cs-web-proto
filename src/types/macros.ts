import React from "react";

/* A simple dictionary from key to value. */
export interface MacroMap {
  [key: string]: string;
}

/* The MacroContext, used for providing macros to children of
   container widgets.

   Descendants can update the macros of the nearest provider using
   the updateMacro() function.
*/
export interface MacroContextType {
  macros: MacroMap;
  updateMacro: (key: string, value: string) => void;
}
export const MacroContext = React.createContext<MacroContextType>({
  macros: {},
  updateMacro: (key: string, value: string) => {}
});

function interpolate(
  str: string,
  substitutions: { [substitution: string]: string }
): string {
  const requiredMappings: string[] = [];

  const regexp = /\$[{(](.*?)[})]/g;
  let result = null;
  while ((result = regexp.exec(str))) {
    requiredMappings.push(result[1]);
  }
  const missingMappings = requiredMappings.filter(
    (x: string): boolean => substitutions[x] === undefined
  );

  /* Allow missing macros to go unresolved. */
  for (const missing of missingMappings) {
    substitutions[missing] = "${" + missing + "}";
  }

  return str.replace(/\$[{(](.*?)[})]/g, (x, g): string => substitutions[g]);
}

export function resolveMacros(
  unresolvedPvName: string,
  macroMap: MacroMap
): string {
  const resolved = interpolate(unresolvedPvName, macroMap);
  return resolved;
}
