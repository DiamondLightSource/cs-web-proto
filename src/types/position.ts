import { CSSProperties } from "react";

export type Position = AbsolutePosition | RelativePosition;

function invalidSize(size?: string): boolean {
  return size === "" || size === undefined;
}

export class AbsolutePosition {
  public x: string;
  public y: string;
  public width: string;
  public height: string;
  public margin: string;
  public padding: string;
  public minWidth: string;
  public maxWidth: string;
  public minHeight: string;

  public constructor(
    x: string,
    y: string,
    width: string,
    height: string,
    margin = "",
    padding = "",
    minWidth = "",
    maxWidth = "",
    minHeight = ""
  ) {
    if (
      invalidSize(x) ||
      invalidSize(y) ||
      invalidSize(width) ||
      invalidSize(height)
    ) {
      throw new Error(
        `Invalid AbsolutePosition (${x},${y},${width},${height})`
      );
    }
    this.x = x;
    this.y = y;
    this.width = width;
    this.height = height;
    this.margin = margin;
    this.padding = padding;
    this.minWidth = minWidth;
    this.maxWidth = maxWidth;
    this.minHeight = minHeight;
  }

  public css(): CSSProperties {
    return {
      position: "absolute",
      top: this.y,
      left: this.x,
      width: this.width,
      height: this.height,
      margin: this.margin,
      padding: this.padding,
      minWidth: this.minWidth,
      maxWidth: this.maxWidth,
      minHeight: this.minHeight
    };
  }

  public toString(): string {
    return `AbsolutePosition (${this.x},${this.y},${this.width},${this.height})`;
  }
}

export class RelativePosition {
  public width: string;
  public height: string;
  public margin: string;
  public padding: string;
  public minWidth: string;
  public maxWidth: string;
  public minHeight: string;

  public constructor(
    width = "",
    height = "",
    margin = "",
    padding = "",
    minWidth = "",
    maxWidth = "",
    minHeight = ""
  ) {
    this.width = width;
    this.height = height;
    this.margin = margin;
    this.padding = padding;
    this.minWidth = minWidth;
    this.maxWidth = maxWidth;
    this.minHeight = minHeight;
  }

  public css(): CSSProperties {
    return {
      position: "relative",
      width: this.width,
      height: this.height,
      margin: this.margin,
      padding: this.padding,
      minWidth: this.minWidth,
      maxWidth: this.maxWidth,
      minHeight: this.minHeight
    };
  }

  public toString(): string {
    return `RelativePosition (${this.width},${this.height})`;
  }
}
