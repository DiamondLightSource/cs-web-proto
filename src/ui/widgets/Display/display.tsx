import React from "react";

import { Widget } from "../widget";
import { Component, WidgetPropType } from "../widgetProps";
import { registerWidget } from "../register";
import { ChoicePropOpt, ChildrenPropOpt, InferWidgetProps } from "../propTypes";

const DisplayProps = {
  children: ChildrenPropOpt,
  overflow: ChoicePropOpt(["scroll", "hidden", "auto", "visible"])
};

// Generic display widget to put other things inside
const DisplayComponent = (
  props: InferWidgetProps<typeof DisplayProps> & Component
): JSX.Element => (
  <div
    style={{
      position: "relative",
      boxSizing: "border-box",
      overflow: props.overflow,
      ...props.style
    }}
  >
    {props.children}
  </div>
);

const DisplayWidgetProps = {
  ...DisplayProps,
  ...WidgetPropType
};

export const Display = (
  props: InferWidgetProps<typeof DisplayWidgetProps>
): JSX.Element => <Widget baseWidget={DisplayComponent} {...props} />;

registerWidget(Display, DisplayWidgetProps, "display");
