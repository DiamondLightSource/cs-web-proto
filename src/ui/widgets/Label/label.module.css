.Label {
  left: 0%;
  top: 0%;
  height: 100%;
  width: 100%;
  display: flex;
  padding: 3px;
  background-color: #888888;
  font: 14px helvetica, sans-serif;
  text-align: center;
  justify-content: center;
  align-items: center;
}
