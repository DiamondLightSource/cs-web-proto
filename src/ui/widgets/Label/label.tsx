import React from "react";

import classes from "./label.module.css";
import { Widget } from "../widget";
import { WidgetPropType } from "../widgetProps";
import { registerWidget } from "../register";
import {
  BoolPropOpt,
  StringProp,
  InferWidgetProps,
  StringPropOpt,
  ChoicePropOpt,
  FontPropOpt,
  ColorPropOpt,
  BorderPropOpt
} from "../propTypes";

const LabelProps = {
  text: StringProp,
  visible: BoolPropOpt,
  transparent: BoolPropOpt,
  className: StringPropOpt,
  textAlign: ChoicePropOpt(["left", "center", "right"]),
  font: FontPropOpt,
  foregroundColor: ColorPropOpt,
  backgroundColor: ColorPropOpt,
  border: BorderPropOpt
};

const LabelWidgetProps = {
  ...LabelProps,
  ...WidgetPropType
};

type LabelWidget = InferWidgetProps<typeof LabelWidgetProps>;

export const LabelComponent = (
  props: InferWidgetProps<typeof LabelProps>
): JSX.Element => {
  const { visible = true, transparent = false, textAlign = "center" } = props;
  const className = props.className ?? `Label ${classes.Label}`;
  const style: any = { ...props.font?.asStyle(), ...props.border?.asStyle() };
  if (!visible) {
    style["visibility"] = "hidden";
  }
  // Since display is "flex", use "flex-start" and "flex-end" to align
  // the content.
  let alignment = "center";
  if (textAlign === "left") {
    alignment = "flex-start";
  } else if (textAlign === "right") {
    alignment = "flex-end";
  }
  style["justifyContent"] = alignment;
  style["color"] = props.foregroundColor?.rgbaString();
  style["backgroundColor"] = props.backgroundColor?.rgbaString();
  // Transparent prop overrides backgroundColor.
  if (transparent) {
    style["backgroundColor"] = "transparent";
  }
  // Simple component to display text - defaults to black text and dark grey background
  return (
    <div className={className} style={style}>
      {props.text}
    </div>
  );
};

export const Label = (props: LabelWidget): JSX.Element => (
  <Widget baseWidget={LabelComponent} {...props} />
);

registerWidget(Label, LabelWidgetProps, "label");
