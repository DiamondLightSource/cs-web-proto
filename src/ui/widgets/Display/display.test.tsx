import React from "react";
import { shallow, ShallowWrapper } from "enzyme";
import { create, ReactTestRenderer } from "react-test-renderer";
import { DisplayComponent } from "./display";
import { Label } from "../Label/label";
import { RelativePosition } from "../../../types/position";
import { MacroMap } from "../../../types/macros";

let snapshot: ReactTestRenderer;
let wrapper: ShallowWrapper;

// Mock useSelector to return a 'global' macro map.
jest.mock("react-redux", (): object => {
  return {
    useSelector: (): MacroMap => ({
      A: "B",
      C: "D"
    }),
    useDispatch: (): void => {}
  };
});

beforeEach((): void => {
  const display = (
    <DisplayComponent id="id1">
      <Label text="hello" position={new RelativePosition()} />
    </DisplayComponent>
  );
  wrapper = shallow(display);
  snapshot = create(display);
});

describe("<Display />", (): void => {
  test("it matches the snapshot", (): void => {
    expect(snapshot.toJSON()).toMatchSnapshot();
  });

  test("it renders a div", (): void => {
    expect(wrapper.childAt(0).type()).toEqual("div");
  });
});
