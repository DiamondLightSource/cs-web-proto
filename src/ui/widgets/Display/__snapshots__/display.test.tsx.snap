// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`<Display /> it matches the snapshot 1`] = `
<div
  style={
    Object {
      "backgroundColor": undefined,
      "height": "100%",
      "opacity": undefined,
      "overflow": undefined,
      "position": "relative",
      "visibility": undefined,
    }
  }
>
  <div
    className=""
    onMouseDown={[Function]}
    onMouseUp={[Function]}
    style={
      Object {
        "height": "",
        "margin": "",
        "maxWidth": "",
        "minHeight": "",
        "minWidth": "",
        "padding": "",
        "position": "relative",
        "width": "",
      }
    }
  >
    <div
      className="Label Label"
      style={
        Object {
          "backgroundColor": undefined,
          "color": undefined,
          "justifyContent": "center",
          "opacity": undefined,
          "visibility": undefined,
        }
      }
    >
      hello
    </div>
  </div>
</div>
`;
